curly: multi-or-nest option reports an error for a single-statment block with a semicolon on the next line

**Tell us about your environment**
* **ESLint Version:** 6.5.1
* **Node Version:** 12.11.0
* **npm Version:** 6.11.3

**What parser (default, Babel-ESLint, etc.) are you using?** default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```json
{
  "rules": {
    "curly": [2, "multi-or-nest"]
  }
}
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

I used the following code in `file.js`:
```js
if (true) console.log('some statement')
;(function () {})()
```

The issue occurs whenever the statement in the if block (or any block) has the semicolon on the following line.

Then I linted it:
```bash
eslint file.js
```

**What did you expect to happen?**
No error would occur as there is only 1 statement in the if block.

**What actually happened? Please include the actual, raw output from ESLint.**
I got the following error:
```
file.js
  1:1  error  Expected { after 'if' condition  curly

✖ 1 problem (1 error, 0 warnings)
  1 error and 0 warnings potentially fixable with the `--fix` option.
```

When fixing it with `eslint --fix file.js`, it gets changed to
```js
if (true) {console.log('some statement')
;}(function () {})()
```
which conflicts with the `"semi": [2, "never"]` rule as that would change it to:
```js
if (true) {console.log('some statement')
}(function () {})()
```
which would cause `curly` to report `Unnecessary { after 'if' condition`.

**Are you willing to submit a pull request to fix this bug?**
Yes.

--- a/lib/rules/curly.js
+++ b/lib/rules/curly.js
@@ -97,10 +97,15 @@ module.exports = {
          * @private
          */
         function isOneLiner(node) {
-            const first = sourceCode.getFirstToken(node),
-                last = sourceCode.getLastToken(node);
+            if (node.type === "EmptyStatement") {
+                return true;
+            }
 
-            return first.loc.start.line === last.loc.end.line;
+            const first = sourceCode.getFirstToken(node);
+            const last = sourceCode.getLastToken(node);
+            const lastExcludingSemicolon = astUtils.isSemicolonToken(last) ? sourceCode.getTokenBefore(last) : last;
+
+            return first.loc.start.line === lastExcludingSemicolon.loc.end.line;
         }
 
         /**

--- a/tests/lib/rules/curly.js
+++ b/tests/lib/rules/curly.js
@@ -216,11 +216,48 @@ ruleTester.run("curly", rule, {
             parserOptions: { ecmaVersion: 6 }
         },
         {
-            code: "while (true) doSomething() \n ;",
+            code: "while (foo) doSomething() \n ;",
             options: ["multi-or-nest"]
         },
         {
             code: "do doSomething() \n ;while (foo)",
+            options: ["multi-or-nest"]
+        },
+        {
+            code: "if (foo)\n;",
+            options: ["multi-or-nest"]
+        },
+        {
+            code: "if (foo) doSomething(); \n else if (bar)\n;",
+            options: ["multi-or-nest"]
+        },
+        {
+            code: "if (foo) doSomething(); \n else\n;",
+            options: ["multi-or-nest"]
+        },
+        {
+            code: "if (foo) doSomething(); \n else if (bar) doSomethingElse(); \n else\n;",
+            options: ["multi-or-nest"]
+        },
+        {
+            code: "for (var i = 0; foo; i++)\n;",
+            options: ["multi-or-nest"]
+        },
+        {
+            code: "for (var foo in bar)\n;",
+            options: ["multi-or-nest"]
+        },
+        {
+            code: "for (var foo of bar)\n;",
+            options: ["multi-or-nest"],
+            parserOptions: { ecmaVersion: 6 }
+        },
+        {
+            code: "while (foo)\n;",
+            options: ["multi-or-nest"]
+        },
+        {
+            code: "do\n;while (foo)",
             options: ["multi-or-nest"]
         },
 
@@ -936,6 +973,57 @@ ruleTester.run("curly", rule, {
             output: "if (true)\n{foo()\n;}[1, 2, 3].bar()",
             options: ["multi-line"],
             errors: [{ messageId: "missingCurlyAfterCondition", data: { name: "if" }, type: "IfStatement" }]
+        },
+
+        // https://github.com/eslint/eslint/issues/12370
+        {
+            code: "if (foo) {\ndoSomething()\n;\n}",
+            output: "if (foo) \ndoSomething()\n;\n",
+            options: ["multi-or-nest"],
+            errors: [{ messageId: "unexpectedCurlyAfterCondition", data: { name: "if" }, type: "IfStatement" }]
+        },
+        {
+            code: "if (foo) doSomething();\nelse if (bar) {\ndoSomethingElse()\n;\n}",
+            output: "if (foo) doSomething();\nelse if (bar) \ndoSomethingElse()\n;\n",
+            options: ["multi-or-nest"],
+            errors: [{ messageId: "unexpectedCurlyAfterCondition", data: { name: "if" }, type: "IfStatement" }]
+        },
+        {
+            code: "if (foo) doSomething();\nelse {\ndoSomethingElse()\n;\n}",
+            output: "if (foo) doSomething();\nelse \ndoSomethingElse()\n;\n",
+            options: ["multi-or-nest"],
+            errors: [{ messageId: "unexpectedCurlyAfter", data: { name: "else" }, type: "IfStatement" }]
+        },
+        {
+            code: "for (var i = 0; foo; i++) {\ndoSomething()\n;\n}",
+            output: "for (var i = 0; foo; i++) \ndoSomething()\n;\n",
+            options: ["multi-or-nest"],
+            errors: [{ messageId: "unexpectedCurlyAfterCondition", data: { name: "for" }, type: "ForStatement" }]
+        },
+        {
+            code: "for (var foo in bar) {\ndoSomething()\n;\n}",
+            output: "for (var foo in bar) \ndoSomething()\n;\n",
+            options: ["multi-or-nest"],
+            errors: [{ messageId: "unexpectedCurlyAfter", data: { name: "for-in" }, type: "ForInStatement" }]
+        },
+        {
+            code: "for (var foo of bar) {\ndoSomething()\n;\n}",
+            output: "for (var foo of bar) \ndoSomething()\n;\n",
+            options: ["multi-or-nest"],
+            parserOptions: { ecmaVersion: 6 },
+            errors: [{ messageId: "unexpectedCurlyAfter", data: { name: "for-of" }, type: "ForOfStatement" }]
+        },
+        {
+            code: "while (foo) {\ndoSomething()\n;\n}",
+            output: "while (foo) \ndoSomething()\n;\n",
+            options: ["multi-or-nest"],
+            errors: [{ messageId: "unexpectedCurlyAfterCondition", data: { name: "while" }, type: "WhileStatement" }]
+        },
+        {
+            code: "do {\ndoSomething()\n;\n} while (foo)",
+            output: "do \ndoSomething()\n;\n while (foo)",
+            options: ["multi-or-nest"],
+            errors: [{ messageId: "unexpectedCurlyAfter", data: { name: "do" }, type: "DoWhileStatement" }]
         }
     ]
 });

--- a/tests/lib/rules/curly.js
+++ b/tests/lib/rules/curly.js
@@ -185,6 +185,45 @@ ruleTester.run("curly", rule, {
             parserOptions: { ecmaVersion: 6 }
         },
 
+        // https://github.com/eslint/eslint/issues/12370
+        {
+            code: "if (foo) doSomething() \n ;",
+            options: ["multi-or-nest"]
+        },
+        {
+            code: "if (foo) doSomething(); \n else if (bar) doSomethingElse() \n ;",
+            options: ["multi-or-nest"]
+        },
+        {
+            code: "if (foo) doSomething(); \n else doSomethingElse() \n ;",
+            options: ["multi-or-nest"]
+        },
+        {
+            code: "if (foo) doSomething(); \n else if (bar) doSomethingElse(); \n else doAnotherThing() \n ;",
+            options: ["multi-or-nest"]
+        },
+        {
+            code: "for (var i = 0; foo; i++) doSomething() \n ;",
+            options: ["multi-or-nest"]
+        },
+        {
+            code: "for (var foo in bar) console.log(foo) \n ;",
+            options: ["multi-or-nest"]
+        },
+        {
+            code: "for (var foo of bar) console.log(foo) \n ;",
+            options: ["multi-or-nest"],
+            parserOptions: { ecmaVersion: 6 }
+        },
+        {
+            code: "while (true) doSomething() \n ;",
+            options: ["multi-or-nest"]
+        },
+        {
+            code: "do doSomething() \n ;while (foo)",
+            options: ["multi-or-nest"]
+        },
+
         // https://github.com/eslint/eslint/issues/3856
         {
             code: "if (true) { if (false) console.log(1) } else console.log(2)",

